apps/dev/src/commands/oracles/add.ts

```
@@ -0,0 +1,318 @@
+          writeFile(join(targetDir, 'Cargo.toml'), cargoToml, 'utf8'),
+        ),
+        Effect.tryPromise(() =>
+          writeFile(join(targetDir, 'spin.toml'), spinToml, 'utf8'),
+        ),
+        Effect.tryPromise(() =>
+          writeFile(join(targetDir, 'src', 'lib.rs'), libRs, 'utf8'),
+        ),
+      ],
+      { concurrency: 'unbounded' },
+    );
+  });
+
+  return Effect.tapError(build, error =>
+    Effect.flatMap(
+      Effect.sync(() => {
+        console.error(
+          `[oracle:add] failed to scaffold workspace '${ctx.workspaceName}': ` +
+            `${error instanceof Error ? error.message : String(error)} — removing partial directory ${targetDir}`,
+        );
+      }),
+      () =>
+        Effect.tryPromise(() =>
+          rm(targetDir, { recursive: true, force: true }),
+        ),
+    ),
+  );
+}
+
+function renderTemplate(templateRelPath: string, context: TemplateContext) {
+  return Effect.gen(function* () {
+    const tpl = (yield* Effect.tryPromise(() =>
+      readFile(join(DEV_TEMPLATES_DIR, templateRelPath), 'utf8'),
+    )) as string;
+    const rendered = yield* Effect.try({
+      try: () => ejs.render(tpl, context, { rmWhitespace: false }),
+      catch: e => e as Error,
+    });
+    return rendered as string;
+  });
+}
+
+function addToCargoMembers(workspaceName: string) {
+  return Effect.gen(function* () {
+    const filePath = join(ORACLES_DIR, 'Cargo.toml');
+    const original = (yield* Effect.tryPromise(() =>
+      readFile(filePath, 'utf8'),
+    )) as string;
+    const parsed = parseToml(original) as any;
+
+    if (!parsed.workspace) parsed.workspace = {};
+    if (!Array.isArray(parsed.workspace.members)) parsed.workspace.members = [];
+    const members: string[] = parsed.workspace.members;
+    if (!members.includes(workspaceName)) members.push(workspaceName);
+
+    const wsStart = original.indexOf('[workspace]');
+    if (wsStart === -1) return; // silently skip if missing
```

Consider printing a warning before returning. ( Maybe use `Effect.logWarning` )
This can be applied to all places where we 'silently skip'